text/0022-pibd-deployment.md

```diff
@@ -0,0 +1,444 @@
+
+```
+BITMAP_SEGMENT_HEIGHT = 9
+OUTPUT_SEGMENT_HEIGHT = 11
+RANGEPROOF_SEGMENT_HEIGHT = 11
+KERNEL_SEGMENT_HEIGHT = 11
+```
+The processes outlined in later sections will assume constant segment heights, but implementations may choose to adjust these heights and/or perform segment size calculations dynamically.
+
+To determine the number of segments `n` of height `h` required for an MMR of a given size `s`, including a possible incomplete last segment, the following calculation can be used:
+
+```
+ceiling(n_leaves(s) / pow(2, h))
+```
+
+where `n_leaves` is the function that calculates the number of leaves in a pmmr of a given size.
+
+Alternatively, a node can attempt to blindly request segments sequentually until one is returned that is not a 'full' segment, however this strategy may lead to unnecessary requests and difficulties determining how much of the PIBD process remains.
+
+#### Peer Selection
+
+The syncing node should select a different peer for each segment request according to the following strategy:
+
+1. Begin with a list of all known peers
+1. Filter this list to only include peers advertising the maximum known difficulty on the network 
+1. Further filter this list to only include peers advertising the [PIBD_HIST_1](#capabilities-flag) flag
+1. Filter this list to exclude peers that have not responded to segment requests recently
+1. Select a peer at random from the remaining list of peers
+
+For the deployment period between [the initial rollout of PIBD and the retirement of the `txhashset.zip` sync method](#rollout-timing), nodes should abort the PIBD process and fall back to the `txhashset.zip` method of syncing if the following are true:
+
+* There are no peers advertising the maximum known difficulty on the network also advertising the `PIBD_HIST_1` capability
+* The node has not seen such a peer for 60 seconds
+
+#### Requesting output bitmap segments
+
+Before any other segments can be validated or applied to local MMR storage, the syncing node must first request and validate the complete output bitmap set up to the MMR size represented in the horizon header.
+
+The MMR size, number of leaves and required number of bitmap segments can be pre-determined based on the size of the output MMR in the horizon header as demonstrated by the following reference code snippet:
+
+```
+// Calculate number of leaves in the bitmap mmr
+bitmap_mmr_leaf_count = (pmmr::n_leaves(self.archive_header.output_mmr_size) + 1023) / 1024;
+
+// Total size of Bitmap PMMR
+bitmap_mmr_size = 1 + pmmr::peaks(pmmr::insertion_to_pmmr_index(self.bitmap_mmr_leaf_count)).last()
```

Comment:
Thought I mentioned this simplification before?!

bitmap_mmr_size = pmmr::insertion_to_pmmr_index(self.bitmap_mmr_leaf_count)

Reply:
noted and updated